src/porepy/numerics/interface_laws/contact_mechanics_interface_laws.py

```diff
@@ -0,0 +1,781 @@
+            raise ValueError(
+                """The number of dofs of the slave discretization given
+            in RobinCoupling must match the number of dofs given by the matrix
+            """
+            )
+        elif not mg.num_cells * ambient_dimension == matrix[master_ind, 2].shape[1]:
+            raise ValueError(
+                """The number of dofs of the edge discretization given
+            in the PrimalContactCoupling must match the number of dofs given by the matrix
+            """
+            )
+
+        # We know the number of dofs from the master and slave side from their
+        # discretizations
+        dof = np.array(
+            [
+                matrix[master_ind, master_ind].shape[1],
+                matrix[slave_ind, slave_ind].shape[1],
+                mg.num_cells * ambient_dimension,
+            ]
+        )
+        cc = np.array([sps.coo_matrix((i, j)) for i in dof for j in dof])
+        cc = cc.reshape((3, 3))
+
+        rhs = np.empty(3, dtype=np.object)
+        rhs[master_ind] = np.zeros(dof_master)
+        rhs[slave_ind] = np.zeros(dof_slave)
+        rhs[mortar_ind] = np.zeros(mg.num_cells * ambient_dimension)
+
+        master_scalar_gradient = data_master[pp.DISCRETIZATION_MATRICES][self.keyword][
+            "grad_p"
+        ]
+
+        # We want to modify the stress balance posed on the edge to account for the
+        # scalar (usually pressure) contribution.
+        # In the purely mechanical case, stress from the higher dimensional
+        # domain (both interior and bound_stress) should match the contact stress:
+        # -\lambda_slave + \lambda_master = 0,
+        # see PrimalContactCoupling.
+        # The following modification is needed:
+        # Add the scalar gradient contribution to the traction on the master
+        # boundary.
+
+        # A diagonal operator is needed to switch the sign of vectors on
+        # higher-dimensional faces that point into the fracture surface, see
+        # PrimalContactCoupling.
+        faces_on_fracture_surface = mg.master_to_mortar_int().tocsr().indices
+        sign_switcher = pp.grid_utils.switch_sign_if_inwards_normal(
+            g_master, ambient_dimension, faces_on_fracture_surface
+        )
+
+        # i) Obtain pressure stress contribution from the higher dimensional domain.
+        # ii) Switch the direction of the vectors, so that for all faces, a positive
+        # force points into the fracture surface (along the outwards normal on the
+        # boundary).
+        # iii) Map to the mortar grid.
+        # iv) Minus according to - alpha grad p
+        master_scalar_stress_to_master_traction = -(
+            mg.master_to_mortar_int(nd=ambient_dimension)
+            * sign_switcher
+            * master_scalar_gradient
+        )
+        cc[mortar_ind, master_ind] = master_scalar_stress_to_master_traction
+
+        matrix += cc
+
+        return matrix, rhs
+
+
+class FractureScalarToForceBalance:
```

Comment:
I think it would be better to merge the FractureScalarToForceBalance and MatrixScalarToForceBalance into one discretization. It should be straight forward, and I think it is more consistent.

Further, this will help @keileg when he decides to implement mixed-dimensional elasticity.

Reply:
This might require a change to the handling of the difference between the fracture contribution from temperature and pressure. Perhaps the alpha=0 in the fracture might be an option. Opinions, @keileg?

Reply:
To be reconsidered in another PR. Mixed-dimensional elasticity is coming up in the next months, or so I hope.